src/System.Private.CoreLib/shared/System/DateTime.Windows.cs

```diff
@@ -120,19 +109,101 @@ internal FullSystemTime(long ticks)
             }
         };
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal static extern bool ValidateSystemTime(in Interop.Kernel32.SYSTEMTIME time, bool localTime);
+#if !CORECLR
+        internal static readonly bool s_systemSupportsPreciseSystemTime = SystemSupportsPreciseSystemTime();
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal static extern bool FileTimeToSystemTime(long fileTime, out FullSystemTime time);
+        private static unsafe bool SystemSupportsPreciseSystemTime()
+        {
+            if (Environment.IsWindows8OrAbove)
+            {
+                // GetSystemTimePreciseAsFileTime exists and we'd like to use it.  However, on
+                // misconfigured systems, it's possible for the "precise" time to be inaccurate:
+                //     https://github.com/dotnet/coreclr/issues/14187
+                // If it's inaccurate, though, we expect it to be wildly inaccurate, so as a
+                // workaround/heuristic, we get both the "normal" and "precise" times, and as
+                // long as they're close, we use the precise one. This workaround can be removed
+                // when we better understand what's causing the drift and the issue is no longer
+                // a problem or can be better worked around on all targeted OSes.
+
+                long systemTimeResult;
+                Interop.Kernel32.GetSystemTimeAsFileTime(&systemTimeResult);
+
+                long preciseSystemTimeResult;
+                Interop.Kernel32.GetSystemTimePreciseAsFileTime(&preciseSystemTimeResult);
+
+                return Math.Abs(preciseSystemTimeResult - systemTimeResult) <= 100 * TicksPerMillisecond;
+            }
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal static extern void GetSystemTimeWithLeapSecondsHandling(out FullSystemTime time);
+            return false;
+        }
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal static extern bool SystemTimeToFileTime(in Interop.Kernel32.SYSTEMTIME time, out long fileTime);
+        private static unsafe bool ValidateSystemTime(in Interop.Kernel32.SYSTEMTIME time, bool localTime)
+        {
+            if (localTime)
+            {
+                return Interop.Kernel32.TzSpecificLocalTimeToSystemTime(IntPtr.Zero, in time, out Interop.Kernel32.SYSTEMTIME st);
+            }
+            else
+            {
+                return Interop.Kernel32.SystemTimeToFileTime(in time, out long timestamp);
+            }
+        }
 
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal static extern long GetSystemTimeAsFileTime();
+        private static bool FileTimeToSystemTime(long fileTime, out FullSystemTime time)
+        {
+            time = new FullSystemTime();
```

Comment:
It would be cheaper to clear just `hundredNanoSecond` here

Reply:
I am not sure I understand the suggestion.

Reply:
Ok, I have missed that this is really just to make the C# compiler happy.